Appending struct to interface array fails to compile - error: field expected

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V full version: V 0.2.4 8cc79e4.8f765ed** 
**OS: linux, Linux Mint 20.3**

<!-- Please include the standalone minimal reproducible code.
If not, it may be closed by the administrators. -->
**What did you do?**

```vlang
struct Vec{
	mut:
		x f64
		y f64
		z f64
}

struct Plane{
	position Vec
	normal Vec
}

struct Sphere{
	position Vec
	radius f64
}

interface Object {
	position Vec
//	intersect(r Ray) (bool,f64,f64)
//	commenting this out does not affect results
}

fn main(){
	mut scene := []Object{}

	scene << Plane{
		position: Vec{0,-10,0},
		normal:   Vec{0,-1,0}
	}
	scene << Sphere{
		position: Vec{0,0,-20},
		radius:   7
	}
}

// v run file.v
```
**What did you see instead?**
```
==================
/tmp/v_1000/repro.10209358551997774371.tmp.c:11642: error: field expected
...
==================
(Use `v -cg` to print the entire error message)

builder error:
...
```

Appending struct to interface array fails to compile!

--- a/vlib/v/gen/c/struct.v
+++ b/vlib/v/gen/c/struct.v
@@ -183,6 +183,9 @@ fn (mut g Gen) struct_init(node ast.StructInit) {
 					}
 				}
 				if !cloned {
+					inside_cast_in_heap := g.inside_cast_in_heap
+					g.inside_cast_in_heap = 0 // prevent use of pointers in child structs
+
 					if field_type_sym.kind == .array_fixed && sfield.expr is ast.Ident {
 						fixed_array_info := field_type_sym.info as ast.ArrayFixed
 						g.write('{')
@@ -202,6 +205,7 @@ fn (mut g Gen) struct_init(node ast.StructInit) {
 						}
 						g.expr_with_cast(sfield.expr, sfield.typ, sfield.expected_type)
 					}
+					g.inside_cast_in_heap = inside_cast_in_heap // restore value for further struct inits
 				}
 				if is_multiline {
 					g.writeln(',')

--- a/vlib/v/tests/append_struct_to_interface_array_test.v
+++ b/vlib/v/tests/append_struct_to_interface_array_test.v
@@ -1,0 +1,43 @@
+struct Vec {
+mut:
+	x f64
+	y f64
+	z f64
+}
+
+struct Plane {
+	position Vec
+	normal   Vec
+}
+
+struct Sphere {
+	position Vec
+	radius   f64
+}
+
+interface Object {
+	position Vec
+}
+
+fn test_append_struct_to_interface_array() {
+	mut scene := []Object{}
+
+	scene << Plane{
+		position: Vec{0, -10, 0}
+		normal: Vec{0, -1, 0}
+	}
+	scene << Sphere{
+		position: Vec{0, 0, -20}
+		radius: 7
+	}
+
+	println(scene)
+
+	assert scene.len == 2
+	assert scene[0].position.x == 0
+	assert scene[0].position.y == -10
+	assert scene[0].position.z == 0
+	assert scene[1].position.x == 0
+	assert scene[1].position.y == 0
+	assert scene[1].position.z == -20
+}

